File: ActiveRealmSample/ActiveRealmSample/Model/UserSettings.m
//
// Created by Masaki Ando on 2019-06-05.
// Copyright (c) 2019 Hituzi Ando. All rights reserved.
//

#import "UserSettings.h"

@interface ActiveRealmUserSettings : ARMObject

@property NSString *authorID;
@property NSNumber <RLMBool> *notificationEnabled;

@end

RLM_ARRAY_TYPE(ActiveRealmUserSettings)

@implementation ActiveRealmUserSettings

@end

@implementation UserSettings

- (instancetype)init {
    if (self = [super init]) {
        _notificationEnabled = @NO;
    }

    return self;
}

+ (NSArray<NSString *> *)ignoredProperties {
    return @[ @"isNotificationEnabled" ];
}

+ (NSDictionary<NSString *, ARMRelationship *> *)definedRelationships {
    return @{
        @"author": [ARMInverseRelationship inverseRelationshipWithClass:NSClassFromString(@"Author")
                                                                   type:ARMInverseRelationshipTypeBelongsTo]
    };
}

+ (BOOL)validateBeforeSaving:(id)obj {
    UserSettings *userSettings = obj;

    return userSettings.authorID.length > 0 && userSettings.notificationEnabled != nil;
}

- (NSString *)description {
    return [self dictionaryWithValuesForKeys:@[
        @"uid",
        @"authorID",
        @"notificationEnabled",
        @"createdAt",
        @"updatedAt"
    ]].description;
}

#pragma mark - property

- (id)author {
    return self.relations[@"author"].object;
}

- (BOOL)isNotificationEnabled {
    return self.notificationEnabled.boolValue;
}

@end

File: ActiveRealmSwiftSample/ActiveRealmSwiftSample/Model/UserSettings.swift
//
// Created by Masaki Ando on 2019-06-05.
// Copyright (c) 2019 Hituzi Ando. All rights reserved.
//

import Foundation

import ActiveRealm

class ActiveRealmUserSettings: ARMObject {

    @objc dynamic var authorID                      = ""
    @objc dynamic var notificationEnabled: NSNumber = false
}

class UserSettings: ARMActiveRealm {

    @objc var authorID                      = ""
    @objc var notificationEnabled: NSNumber = false

    // A relation property. This property is just alias.

    var author: Author {
        return relations["author"]?.object as! Author
    }

    // A property ignored by ActiveRealm.

    var isNotificationEnabled: Bool {
        return notificationEnabled.boolValue
    }

    override class func ignoredProperties() -> [String] {
        return ["isNotificationEnabled"]
    }

    override class func definedRelationships() -> [String: ARMRelationship] {
        return ["author": ARMInverseRelationship(with: Author.self, type: .belongsTo)]
    }

    override class func validateBeforeSaving(_ obj: Any) -> Bool {
        let userSettings = obj as! UserSettings

        return !userSettings.authorID.isEmpty
    }

    override var description: String {
        let dict = dictionaryWithValues(forKeys: [
            "uid",
            "authorID",
            "notificationEnabled",
            "createdAt",
            "updatedAt"
        ])

        return "\(dict)"
    }
}
